var/spack/repos/builtin/packages/python/package.py

```diff
@@ -208,6 +208,9 @@ class Python(AutotoolsPackage):
     # https://github.com/python/cpython/pull/16717
     patch('intel-3.6.7.patch', when='@3.6.7:3.6.8,3.7.1:3.7.5 %intel')
 
+    # Fixes RPATH issue with Intel compilers
+    patch('python-intel-rpath.patch', when='%intel')
```

Comment:
Does this patch apply correctly to all versions of Python?

Reply:
Sorry for the delay. I was trying to get a little more background on the reason for the change. We first noticed this issue when installing py-h5py with intel 18:

`spack install py-h5py~mpi %intel@18.0.5.8.4.0 ^hdf5~mpi %intel@18.0.5.8.4.0 ^python@3.8:3.8.99 %intel@18.0.5.8.4.0`

[py-h5py.spack-build-out.txt](https://github.com/spack/spack/files/5759239/py-h5py.spack-build-out.txt)

I've attached the output, but the gist of the error is:

```
Traceback (most recent call last):
  File "/home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/intel-18.0.5.8.4.0/py-numpy-1.19.4-azjwmdjf/lib/python3.8/site-packages/numpy/core/__init__.py", line 22, in <module>
    from . import multiarray
  File "/home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/intel-18.0.5.8.4.0/py-numpy-1.19.4-azjwmdjf/lib/python3.8/site-packages/numpy/core/multiarray.py", line 12, in <module>
    from . import overrides
  File "/home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/intel-18.0.5.8.4.0/py-numpy-1.19.4-azjwmdjf/lib/python3.8/site-packages/numpy/core/overrides.py", line 7, in <module>
    from numpy.core._multiarray_umath import (
ImportError: libopenblas.so.0: cannot open shared object file: No such file or directory
```

I then looked at _multiarray_umath.cpython-38-x86_64-linux-gnu.so in the py-numpy install and as expected libopenblas.so is not found and the rpath is incorrect:

```
ryan@ryan-VirtualBox:~/Desktop$ ldd ~/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/intel-18.0.5.8.4.0/py-numpy-1.19.4-azjwmdjf/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so 
	linux-vdso.so.1 (0x00007ffe429d3000)
	libopenblas.so.0 => not found
	libimf.so => /home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/gcc-7.5.0/intel-18.0.5-pt2adjau/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libimf.so (0x00007fcb58468000)
	libsvml.so => /home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/gcc-7.5.0/intel-18.0.5-pt2adjau/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libsvml.so (0x00007fcb56b5a000)
	libirng.so => /home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/gcc-7.5.0/intel-18.0.5-pt2adjau/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libirng.so (0x00007fcb567e6000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fcb56697000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fcb5667a000)
	libintlc.so.5 => /home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/gcc-7.5.0/intel-18.0.5-pt2adjau/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libintlc.so.5 (0x00007fcb5640c000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fcb563e9000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fcb561f7000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fcb59023000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fcb561f1000)
```

```
ryan@ryan-VirtualBox:~/Desktop$ readelf -d ~/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/intel-18.0.5.8.4.0/py-numpy-1.19.4-azjwmdjf/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so 

Dynamic section at offset 0x5d5a28 contains 36 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libopenblas.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libimf.so]
 0x0000000000000001 (NEEDED)             Shared library: [libsvml.so]
 0x0000000000000001 (NEEDED)             Shared library: [libirng.so]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libintlc.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x000000000000000f (RPATH)              Library rpath: [/home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/gcc-7.5.0/intel-18.0.5-pt2adjau/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64]
 0x000000000000000c (INIT)               0x2a000
 0x000000000000000d (FINI)               0x4dd340
 0x0000000000000019 (INIT_ARRAY)         0x5d5a20
 0x000000000000001b (INIT_ARRAYSZ)       8 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x5d5a28
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x0000000000000004 (HASH)               0x298
 0x000000006ffffef5 (GNU_HASH)           0x1a58
 0x0000000000000005 (STRTAB)             0x8540
 0x0000000000000006 (SYMTAB)             0x27c8
 0x000000000000000a (STRSZ)              11960 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x5d7000
 0x0000000000000002 (PLTRELSZ)           15384 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x25670
 0x0000000000000007 (RELA)               0xbc78
 0x0000000000000008 (RELASZ)             104952 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000006ffffffe (VERNEED)            0xbbc8
 0x000000006fffffff (VERNEEDNUM)         4
 0x000000006ffffff0 (VERSYM)             0xb3f8
 0x000000006ffffff9 (RELACOUNT)          4133
 0x0000000000000000 (NULL)               0x0
```

[py-numpy.spack-build-out.txt](https://github.com/spack/spack/files/5759238/py-numpy.spack-build-out.txt)

I then looked at the output from the numpy build. which I have attached. The following warning appears when building 3 different extension, including _multiarray_umath:

`icc: command line warning #10006: ignoring unknown option '-R/home/ryan/Desktop/spack_intel/install/linux-ubuntu20.04-x86_64/intel-18.0.5.8.4.0/openblas-0.3.13-qne7frqb/lib'`

The person on my team who made the original fix dug around and made the patch to unixccompiler.py. We noticed the above error when building py-h5py with python 3.8 and 3.9 and with intel 18 and 19. I believe the error also occurs with when building py-pyzmq with Spack v0.16.0

The odd thing is py-h5py builds correctly without the patch with python 3.6 and 3.7 even though the same icc warning appears in the output from the numpy@1.19.4 builds. We were unable to track down what changed between python 3.7 and 3.8 that made the rpath appear to be correct with 3.7.

Do you recommend that I add a `when` argument so the patch is applied with the python and intel versions that we tested?

Reply:
Hmm, has this patch been contributed upstream? If it doesn't cause any problems for other versions/compilers, no need for a `when` arg. Just want to make sure it applies to 2.7, 3.1, 3.9, etc.

Reply:
@adamjstewart The patch applies correctly for python 3.1+, but fails for python 2.7. This patch address the same issue as [python/cray-rpath-3.1.patch](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/python/cray-rpath-3.1.patch) and [python/fj-rpath-3.1.patch](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/python/fj-rpath-3.1.patch), although all three patches address the issue in slightly different ways. I matched the behavior of the other two patches and limited the patch to `%intel`.

My next move is to contribute the change to the Python repo.